### ISSUE TITLE
Transactionally resilient lookup vindexes

### ISSUE THREAD

**Issue by sougou** [vitessio/vitess#3642]
## Problem
Lookup Vindex entries exist in shards or keyspaces that are usually not in the same shard as the main row that owns them. This gives rise to distributed transactions that will not be atomic without using 2PC.

A number of solutions have been discussed to address this problem. This is the proposal that captures the outcome of those discussions.

## High level approach
It is understood that any non-2pc solution will be imperfect. We just want the solution to be good enough.

The general approach to solving this problem is to perform the vindex operations in their own separate autocommitted transactions:

* Upon insert, a vindex entry will be created and committed upfront. If the insert fails, the vindex entry will remain. However, there is no harm because it will point to non-existent data. The outcome of a subsequent read, although less efficient, will return the correct empty result.

* For delete, we have a few options:
  1. Delete the vindex row as autocommit: the downside of this approach is that the delete of the actual row may fail. This will make that row unreachable. Therefore, this approach is not viable.
  2. Delete the vindex row after the delete succeeds (but before commit). This is still a variation of the first case because we end up in the same situation as the first case.
  3. Delete the vindex row after the final commit. This option is hard to implement because the commit is controlled by the application. The delete statement may be part of a bigger transaction. If so, the post-commit action will need to be stored in the session. This approach also has race conditions that can lead to improper data in vindexes: if create and delete are quickly alternated, the post-commit deletes may not follow the same sequence as the original transaction commit order.
  4. Don't delete the row at all: This option is safe, but eventually inefficient. Every delete will leave a lingering entry in the vindex that could end up causing unnecessarily additional lookups.

Given the above options, the 4th one seems to be the most practical. In the future, we could also look at running a garbage collector that cleans up rows that can be safely deleted.

If we don't do deletes, then there is a possibility that a stray vindex row may prevent future inserts with that key. To address this, all Creates should be changed to upserts. The trade-off with upserts is that uniqueness cannot be guaranteed by the vindex. This is because two rows could be inserted one after the other with the same key, and the vindex will end up being updated to only point to the new row. This can be interpreted two ways:
1. For unique vindexes, the application has the responsibility of ensuring uniqueness.
2. These vindexes cannot be unique.

Since Vitess has generally erred on the side of being permissive, we can go with option 1.

If the app doesn't plan on ever deleting rows, then it can run the Create as regular inserts. In that case, uniqueness constraints can be verified.

* For updates, the operation may or may not succeed. However, proactively modifying the row assuming that the operation will succeed will leave an orphaned row if the update fails. Delaying the update till the main row is updated will lead to the same issues as case 3 of delete. So, it's safer to not allow updates at all. The application has the option of performing a delete in a _separate transaction_ and then follow it with an insert. This will result in the same guarantees previously stated.

## Summary
The app has the following three options:
1. Configure a vindex as non-unique with upserts: This will allow you to insert and delete rows. Too many deleted rows may lead to inefficiencies without a garbage collection process.
2. Configure a vindex as unique with upserts: Uniqueness constraint will not be enforced. Hence, this will be the application's responsibility. Garbage collection may also be needed if there are too many deletes.
3. Configure a vindex as unique with inserts: In this mode, deletes are not allowed (or could be ignored). However, non-unique inserts will result in dup key errors as expected.

Updates will not be allowed.

If there are no objections, I'll work on modifying #3600 to match the above specs.

**Comment by demmer** [comment]
I'm a bit missing the high level motivation here -- can you explain why it's better to have two independent transactions as compared to a single one? It seems like the current semantics do a better job of keeping the owned vindex rows in sync with the owner row, and neither approach addresses the intrinsic lack of atomicity with distributed transactions.

Part of why I ask is that one (major) downside I can see to this approach is that it effectively increases the chance of an expected drift between the lookup vindex and the owned table to the point that we would lose the ability to rely on the lookup table as a reliable source for application queries in their own right.

Specifically, one of the ideas I've been thinking about is to use a multi-column owned lookup_hash vindex as effectively a shadow copy of the owned table, albeit sharded by a different key (possibly in a different keyspace altogether). This would allow the application to directly query the lookup table, or (in a future enhancement), the vtgate query planner could determine that a query can efficiently be resolved _just_ by hitting the lookup vindex, rather than first getting all the keyspace_ids and then scattering the queries out.

Now it is certainly the fact that distributed transactions mean that in the event of a shard failure, timeout, etc, it is of course possible in the current implementation for the owned lookup table to drift from the owner table. However when the infrastructure is functioning with no errors, the current implementation semantics are faithful to mysql's atomicity, and all modifications are made to both tables in the correct order to maintain the one-to-one mapping and uniqueness constraints.

I think this is a very desirable property that would be lost if we went down this approach.

Finally, whether or not we adopt this redesign, the owned lookup table can still drift from the owning table, so I think we would be well-served to invest some time in a lookup vindex verification / repair workflow.

**Comment by sougou** [comment]
An alternate way to approach this feature is to ask how a lookup vindex should work if it was implemented using a non-vitess system (like Reddis):

From this perspective, I think the trade-offs make sense. But I have to agree that they're very unsatisfactory.

We can look at changing the vindex protocol to be 2-phase: so, `Create` would be split into `CreatePrepare` and `CreateCommit`. But all 2-phase protocols require a monitoring tool that watches over abandoned transactions to resolve them.

The advantage of doing it outside of the vitess 2-pc framework is that it will be cheaper, and will therefore be attractive for someone who would otherwise prefer best-effort commit.

In other words, I think this is a good start. I'm thinking of implementing only options 1 & 2 listed in the conclusion. I'm hoping that we can instead focus on making the vitess 2-pc as efficient as a best-effort commit. Then all these become non-issues.

**Comment by sougou** [comment]
As for your other requirement about 'index scans', it feels like an orthogonal feature. Relying on something like `lookup_hash` is definitely a hack.

What we'll probably need is for a lookup vindex to map to a column (or columns), and then Vitess will apply that column (or columns') vindex to compute the keyspace id. The restriction, obviously is that the target columns must have a vindex.

**Comment by sougou** [comment]
And I think we can also support safe autocommits for unique lookups if we performed plain inserts instead of upserts. The restriction will be that you cannot update or delete without cleaning up the stray row, which can be done at the application level.

It's not perfect, but we can do the best we can.

**Comment by sougou** [comment]
This is now addressed to the extent possible.
